Doesn't work with `npm:` specifiers

> ✘ [ERROR] [plugin deno] Failed to call 'deno info' on 'file:///.../.../index.js'

This could potentially be solved with an `unstable` flag. I will open a PR when my last one is merged. Please take a look.

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -37,4 +37,4 @@ jobs:
         run: deno lint
 
       - name: Run tests
-        run: deno task test
+        run: deno test -A

--- a/.gitignore
+++ b/.gitignore
@@ -1,2 +1,4 @@
 deno.lock
 dist/
+node_modules/
+dist/

--- a/README.md
+++ b/README.md
@@ -2,6 +2,12 @@
 
 Deno module resolution for `esbuild`.
 
+- Support for `file:`, `https:`, and `data:` specifiers
+- Support for `npm:` specifiers
+- Support for import maps (including embedded into `deno.json`)
+- Native loader using Deno's global cache directory
+- Portable loader that words in environments with limited permissions
+
 ## Example
 
 This example bundles an entrypoint into a single ESM output.
@@ -25,6 +31,15 @@ console.log(result.outputFiles);
 
 esbuild.stop();
 ```
+
+## Limitations
+
+- The `"portable"` loader does not use the Deno module cache, so all remote
+  specifiers are downloaded on every run.
+- When using the `"portable"` loader, all `npm:` dependencies must be
+  pre-downloaded into a local `node_modules/` directory.
+- `npm:` specifiers are not supported on WASM esbuild builds due to FS access
+  limitations (see https://github.com/evanw/esbuild/pull/2968).
 
 ## Documentation
 

--- a/deno.jsonc
+++ b/deno.jsonc
@@ -1,9 +1,0 @@
-{
-  "tasks": {
-    "test": "deno test -A",
-    "update": "deno run --allow-read=./ --allow-net --allow-write=./ https://deno.land/x/deno_outdated@0.2.4/cli.ts"
-  },
-  "imports": {
-    "a": "https://deno.land/x/a/mod.ts"
-  }
-}

--- a/deps.ts
+++ b/deps.ts
@@ -1,12 +1,16 @@
 import type * as esbuild from "https://deno.land/x/esbuild@v0.17.18/mod.d.ts";
 export type { esbuild };
 export {
+  dirname,
   fromFileUrl,
+  join,
   resolve,
   toFileUrl,
-} from "https://deno.land/std@0.185.0/path/mod.ts";
-export { basename, extname } from "https://deno.land/std@0.185.0/path/mod.ts";
-export * as JSONC from "https://deno.land/std@0.185.0/jsonc/mod.ts";
+} from "https://deno.land/std@0.173.0/path/mod.ts";
+export { copy } from "https://deno.land/std@0.173.0/fs/mod.ts";
+export { basename, extname } from "https://deno.land/std@0.173.0/path/mod.ts";
+export * as JSONC from "https://deno.land/std@0.173.0/encoding/jsonc.ts";
+export { encode as base32Encode } from "https://deno.land/std@0.173.0/encoding/base32.ts";
 export {
   resolveImportMap,
   resolveModuleSpecifier,
@@ -16,3 +20,4 @@ export type {
   Scopes,
   SpecifierMap,
 } from "https://deno.land/x/importmap@0.2.1/mod.ts";
+export { DenoDir } from "https://deno.land/x/deno_cache@0.4.1/mod.ts";

--- a/mod.ts
+++ b/mod.ts
@@ -51,6 +51,15 @@ export interface DenoPluginsOptions {
    * determine what import map to use, if any.
    */
   importMapURL?: string;
+  /**
+   * Specify whether to generate and use a local `node_modules` directory when
+   * using the `native` loader. This is equivalent to the `--node-modules-dir`
+   * flag to the Deno executable.
+   *
+   * This option is ignored when using the `portable` loader, as the portable
+   * loader always uses a local `node_modules` directory.
+   */
+  nodeModulesDir?: boolean;
 }
 
 export function denoPlugins(opts: DenoPluginsOptions = {}): esbuild.Plugin[] {

--- a/mod_test.ts
+++ b/mod_test.ts
@@ -48,6 +48,7 @@ test("remote ts", LOADERS, async (esbuild, loader) => {
   });
   assertEquals(res.warnings, []);
   assertEquals(res.errors, []);
+  assertEquals(res.errors, []);
   assertEquals(res.outputFiles.length, 1);
   const output = res.outputFiles[0];
   assertEquals(output.path, "<stdout>");
@@ -63,6 +64,7 @@ test("local ts", LOADERS, async (esbuild, loader) => {
     entryPoints: ["./testdata/mod.ts"],
   });
   assertEquals(res.warnings, []);
+  assertEquals(res.errors, []);
   assertEquals(res.errors, []);
   assertEquals(res.outputFiles.length, 1);
   const output = res.outputFiles[0];
@@ -274,6 +276,119 @@ test("local json", LOADERS, async (esbuild, loader) => {
     },
   });
 });
+
+test(
+  "npm specifiers global resolver - preact",
+  ["native"],
+  async (esbuild, loader) => {
+    if (esbuild === PLATFORMS.wasm) return;
+    const res = await esbuild.build({
+      ...DEFAULT_OPTS,
+      plugins: [...denoPlugins({ loader })],
+      bundle: true,
+      entryPoints: ["./testdata/npm/preact.tsx"],
+    });
+    assertEquals(res.warnings, []);
+    assertEquals(res.errors, []);
+    assertEquals(res.outputFiles.length, 1);
+    const output = res.outputFiles[0];
+    assertEquals(output.path, "<stdout>");
+    assert(!output.text.includes(`npm:`));
+    const dataURL = `data:application/javascript;base64,${btoa(output.text)}`;
+    const { default: html } = await import(dataURL);
+    assertEquals(html, "<div>hello world</div>");
+  },
+);
+
+test(
+  "npm specifiers global resolver - react",
+  ["native"],
+  async (esbuild, loader) => {
+    if (esbuild === PLATFORMS.wasm) return;
+    const res = await esbuild.build({
+      ...DEFAULT_OPTS,
+      plugins: [...denoPlugins({ loader })],
+      bundle: true,
+      entryPoints: ["./testdata/npm/react.tsx"],
+    });
+    assertEquals(res.warnings, []);
+    assertEquals(res.errors, []);
+    assertEquals(res.outputFiles.length, 1);
+    const output = res.outputFiles[0];
+    assertEquals(output.path, "<stdout>");
+    assert(!output.text.includes(`npm:`));
+    const dataURL = `data:application/javascript;base64,${btoa(output.text)}`;
+    const { default: html } = await import(dataURL);
+    assertEquals(html, "<div>hello world</div>");
+  },
+);
+
+test(
+  "npm specifiers local resolver - preact",
+  LOADERS,
+  async (esbuild, loader) => {
+    if (esbuild === PLATFORMS.wasm) return;
+    const entryPoint =
+      new URL("./testdata/npm/preact.tsx", import.meta.url).href;
+    const tmp = Deno.makeTempDirSync();
+    console.log("tmp", tmp);
+    if (loader === "portable") {
+      new Deno.Command(Deno.execPath(), {
+        args: ["cache", "--node-modules-dir", entryPoint],
+        cwd: tmp,
+      }).outputSync();
+    }
+    const res = await esbuild.build({
+      ...DEFAULT_OPTS,
+      plugins: [...denoPlugins({ loader, nodeModulesDir: true })],
+      bundle: true,
+      absWorkingDir: tmp,
+      entryPoints: [entryPoint],
+    });
+    assertEquals(res.warnings, []);
+    assertEquals(res.errors, []);
+    assertEquals(res.outputFiles.length, 1);
+    const output = res.outputFiles[0];
+    assertEquals(output.path, "<stdout>");
+    assert(!output.text.includes(`npm:`));
+    const dataURL = `data:application/javascript;base64,${btoa(output.text)}`;
+    const { default: html } = await import(dataURL);
+    assertEquals(html, "<div>hello world</div>");
+  },
+);
+
+test(
+  "npm specifiers local resolver - react",
+  LOADERS,
+  async (esbuild, loader) => {
+    if (esbuild === PLATFORMS.wasm) return;
+    const tmp = Deno.makeTempDirSync();
+    const entryPoint =
+      new URL("./testdata/npm/preact.tsx", import.meta.url).href;
+    if (loader === "portable") {
+      new Deno.Command(Deno.execPath(), {
+        args: ["cache", "--node-modules-dir", entryPoint],
+        cwd: tmp,
+      }).outputSync();
+    }
+    const res = await esbuild.build({
+      ...DEFAULT_OPTS,
+      plugins: [...denoPlugins({ loader, nodeModulesDir: true })],
+      bundle: true,
+      absWorkingDir: tmp,
+      entryPoints: [entryPoint],
+    });
+    assertEquals(res.warnings, []);
+    assertEquals(res.errors, []);
+    assertEquals(res.outputFiles.length, 1);
+    const output = res.outputFiles[0];
+    assertEquals(output.path, "<stdout>");
+    assert(!output.text.includes(`npm:`));
+    const dataURL = `data:application/javascript;base64,${btoa(output.text)}`;
+    const { default: html } = await import(dataURL);
+    assertEquals(html, "<div>hello world</div>");
+  },
+);
 
 test("remote http redirects are de-duped", LOADERS, async (esbuild, loader) => {
   const res = await esbuild.build({

--- a/src/deno.ts
+++ b/src/deno.ts
@@ -21,6 +21,7 @@ interface InfoOutput {
   roots: string[];
   modules: ModuleEntry[];
   redirects: Record<string, string>;
+  npmPackages: Record<string, NpmPackage>;
 }
 
 export type ModuleEntry =
@@ -64,11 +65,18 @@ export interface ModuleEntryNode extends ModuleEntryBase {
   moduleName: string;
 }
 
+export interface NpmPackage {
+  name: string;
+  version: string;
+  dependencies: string[];
+}
+
 export interface InfoOptions {
   cwd?: string;
   config?: string;
   importMap?: string;
   lock?: string;
+  nodeModulesDir?: boolean;
 }
 
 let tmpDir: string | undefined;
@@ -98,6 +106,9 @@ async function info(
   // } else if (!options.cwd) {
   //   opts.args.push("--no-lock");
   // }
+  if (options.nodeModulesDir) {
+    opts.args.push("--node-modules-dir");
+  }
   if (options.cwd) {
     opts.cwd = options.cwd;
   } else {
@@ -123,6 +134,7 @@ export class InfoCache {
 
   #modules: Map<string, ModuleEntry> = new Map();
   #redirects: Map<string, string> = new Map();
+  #npmPackages: Map<string, NpmPackage> = new Map();
 
   constructor(options: InfoOptions = {}) {
     this.#options = options;
@@ -142,6 +154,10 @@ export class InfoCache {
     return entry;
   }
 
+  getNpmPackage(id: string): NpmPackage | undefined {
+    return this.#npmPackages.get(id);
+  }
+
   #resolve(specifier: string): string {
     return this.#redirects.get(specifier) ?? specifier;
   }
@@ -152,19 +168,23 @@ export class InfoCache {
   }
 
   async #load(specifier: string): Promise<void> {
-    const { modules, redirects } = await info(specifier, {
-      importMap: this.#options.importMap,
-    });
+    const { modules, redirects, npmPackages } = await info(
+      specifier,
+      this.#options,
+    );
     for (const module of modules) {
       this.#modules.set(module.specifier, module);
     }
     for (const [from, to] of Object.entries(redirects)) {
       this.#redirects.set(from, to);
     }
+    for (const [id, npmPackage] of Object.entries(npmPackages)) {
+      this.#npmPackages.set(id, npmPackage);
+    }
 
     specifier = this.#resolve(specifier);
     const entry = this.#modules.get(specifier);
-    if (entry === undefined) {
+    if (entry === undefined && specifier.startsWith("npm:")) {
       // we hit https://github.com/denoland/deno/issues/18043, so we have to
       // perform another load to get the actual data of the redirected specifier
       await this.#load(specifier);

--- a/src/loader_native.ts
+++ b/src/loader_native.ts
@@ -1,11 +1,14 @@
-import { esbuild, fromFileUrl } from "../deps.ts";
+import { base32Encode, DenoDir, esbuild, fromFileUrl, join } from "../deps.ts";
 import * as deno from "./deno.ts";
 import {
   Loader,
   LoaderResolution,
   mapContentType,
   mediaTypeToLoader,
+  parseNpmSpecifier,
 } from "./shared.ts";
+
+let DENO_DIR: DenoDir | undefined;
 
 export interface NativeLoaderOptions {
   infoOptions?: deno.InfoOptions;
@@ -22,8 +25,20 @@ export class NativeLoader implements Loader {
     const entry = await this.#infoCache.get(specifier.href);
     if ("error" in entry) throw new Error(entry.error);
 
-    if (entry.kind === "npm" || entry.kind === "node") {
-      throw new Error("Unsupported module kind: " + entry.kind);
+    if (entry.kind === "npm") {
+      // TODO(lucacasonato): remove parsing once https://github.com/denoland/deno/issues/18043 is resolved
+      const parsed = parseNpmSpecifier(new URL(entry.specifier));
+      return {
+        kind: "npm",
+        packageId: entry.npmPackage,
+        packageName: parsed.name,
+        path: parsed.path ?? "",
+      };
+    } else if (entry.kind === "node") {
+      return {
+        kind: "node",
+        path: entry.specifier,
+      };
     }
 
     return { kind: "esm", specifier: new URL(entry.specifier) };
@@ -53,5 +68,37 @@ export class NativeLoader implements Loader {
       res.watchFiles = [fromFileUrl(specifier)];
     }
     return res;
+  }
+
+  nodeModulesDirForPackage(npmPackageId: string): string {
+    const npmPackage = this.#infoCache.getNpmPackage(npmPackageId);
+    if (!npmPackage) throw new Error("NPM package not found.");
+    if (!DENO_DIR) DENO_DIR = new DenoDir(undefined, true);
+    let name = npmPackage.name;
+    if (name.toLowerCase() !== name) {
+      name = `_${base32Encode(new TextEncoder().encode(name))}`;
+    }
+    return join(
+      DENO_DIR.root,
+      "npm",
+      "registry.npmjs.org",
+      name,
+      npmPackage.version,
+    );
+  }
+
+  packageIdFromNameInPackage(
+    name: string,
+    parentPackageId: string,
+  ): string {
+    const parentPackage = this.#infoCache.getNpmPackage(parentPackageId);
+    if (!parentPackage) throw new Error("NPM package not found.");
+    if (parentPackage.name === name) return parentPackageId;
+    for (const dep of parentPackage.dependencies) {
+      const depPackage = this.#infoCache.getNpmPackage(dep);
+      if (!depPackage) throw new Error("NPM package not found.");
+      if (depPackage.name === name) return dep;
+    }
+    throw new Error("NPM package not found.");
   }
 }

--- a/src/loader_portable.ts
+++ b/src/loader_portable.ts
@@ -5,6 +5,7 @@ import {
   LoaderResolution,
   mapContentType,
   mediaTypeToLoader,
+  parseNpmSpecifier,
 } from "./shared.ts";
 
 interface Module {
@@ -29,6 +30,18 @@ export class PortableLoader implements Loader {
       case "data:": {
         const module = await this.#loadRemote(specifier.href);
         return { kind: "esm", specifier: new URL(module.specifier) };
+      }
+      case "npm:": {
+        const npmSpecifier = parseNpmSpecifier(specifier);
+        return {
+          kind: "npm",
+          packageId: "",
+          packageName: npmSpecifier.name,
+          path: npmSpecifier.path ?? "",
+        };
+      }
+      case "node:": {
+        return { kind: "node", path: specifier.pathname };
       }
       default:
         throw new Error(`Unsupported scheme: '${specifier.protocol}'`);

--- a/src/plugin_deno_loader.ts
+++ b/src/plugin_deno_loader.ts
@@ -1,6 +1,10 @@
-import { esbuild } from "../deps.ts";
+import { esbuild, join } from "../deps.ts";
 import { NativeLoader } from "./loader_native.ts";
 import { PortableLoader } from "./loader_portable.ts";
+import {
+  IN_NODE_MODULES,
+  IN_NODE_MODULES_RESOLVED,
+} from "./plugin_deno_resolver.ts";
 import {
   esbuildResolutionToURL,
   Loader,
@@ -54,6 +58,15 @@ export interface DenoLoaderPluginOptions {
   //  * ESM modules.
   //  */
   // lockPath?: string;
+  /**
+   * Specify whether to generate and use a local `node_modules` directory when
+   * using the `native` loader. This is equivalent to the `--node-modules-dir`
+   * flag to the Deno executable.
+   *
+   * This option is ignored when using the `portable` loader, as the portable
+   * loader always uses a local `node_modules` directory.
+   */
+  nodeModulesDir?: boolean;
 }
 
 const LOADERS = ["native", "portable"] as const;
@@ -81,10 +94,13 @@ export const DEFAULT_LOADER: typeof LOADERS[number] =
  * The native loader shells out to the Deno executable under the hood to load
  * files. Requires `--allow-read` and `--allow-run`. In this mode the download
  * cache is shared with the Deno executable. This mode respects deno.lock,
- * DENO_DIR, DENO_AUTH_TOKENS, and all similar loading configuration.
+ * DENO_DIR, DENO_AUTH_TOKENS, and all similar loading configuration. Files are
+ * cached on disk in the same Deno cache as the Deno executable, and will not be
+ * re-downloaded on subsequent builds.
  *
- * When the plugin is configured to use the native loader, it can auto-discover
- * deno.lock and all similar configuration files if the `cwd` option is set.
+ * NPM specifiers can be used in the native loader without requiring a local
+ * `node_modules` directory. NPM packages are resolved, downloaded, cached, and
+ * loaded in the same way as the Deno executable does.
  *
  * ### Portable Loader
  *
@@ -92,6 +108,10 @@ export const DEFAULT_LOADER: typeof LOADERS[number] =
  * Requires `--allow-read` and/or `--allow-net`. This mode does not respect
  * deno.lock, DENO_DIR, DENO_AUTH_TOKENS, or any other loading configuration. It
  * does not cache downloaded files. It will re-download files on every build.
+ *
+ * NPM specifiers can be used in the portable loader, but require a local
+ * `node_modules` directory. The `node_modules` directory must be created prior
+ * using Deno's `--node-modules-dir` flag.
  */
 export function denoLoaderPlugin(
   options: DenoLoaderPluginOptions = {},
@@ -103,17 +123,29 @@ export function denoLoaderPlugin(
   return {
     name: "deno-loader",
     setup(build) {
+      const cwd = build.initialOptions.absWorkingDir ?? Deno.cwd();
+
+      let nodeModulesDir: string | null = null;
+      if (options.nodeModulesDir) {
+        nodeModulesDir = join(cwd, "node_modules");
+      }
+
       let loaderImpl: Loader;
 
+      const packageIdMapping = new Map<string, string>();
+
       build.onStart(function onStart() {
+        packageIdMapping.clear();
         switch (loader) {
           case "native":
             loaderImpl = new NativeLoader({
               infoOptions: {
+                cwd,
                 config: options.configPath,
                 importMap: options.importMapURL,
                 // TODO(lucacasonato): https://github.com/denoland/deno/issues/18159
                 // lock: options.lockPath,
+                nodeModulesDir: options.nodeModulesDir,
               },
             });
             break;
@@ -122,9 +154,84 @@ export function denoLoaderPlugin(
         }
       });
 
+      async function resolveInNodeModules(
+        path: string,
+        packageId: string,
+        resolveDir: string,
+        kind: esbuild.ImportKind,
+      ): Promise<esbuild.OnResolveResult> {
+        const result = await build.resolve(path, {
+          kind,
+          resolveDir,
+          pluginData: IN_NODE_MODULES_RESOLVED,
+        });
+        result.pluginData = IN_NODE_MODULES;
+        packageIdMapping.set(result.path, packageId);
+        return result;
+      }
+
       async function onResolve(
         args: esbuild.OnResolveArgs,
       ): Promise<esbuild.OnResolveResult | null | undefined> {
+        if (args.namespace === "file" && args.pluginData === IN_NODE_MODULES) {
+          if (nodeModulesDir) {
+            const result = await build.resolve(args.path, {
+              kind: args.kind,
+              resolveDir: args.resolveDir,
+              importer: args.importer,
+              namespace: args.namespace,
+              pluginData: IN_NODE_MODULES_RESOLVED,
+            });
+            result.pluginData = IN_NODE_MODULES;
+            return result;
+          } else if (
+            loaderImpl.nodeModulesDirForPackage &&
+            loaderImpl.packageIdFromNameInPackage
+          ) {
+            const parentPackageId = packageIdMapping.get(args.importer);
+            if (!parentPackageId) {
+              throw new Error(
+                `Could not find package ID for importer: ${args.importer}`,
+              );
+            }
+            if (args.path.startsWith(".")) {
+              return resolveInNodeModules(
+                args.path,
+                parentPackageId,
+                args.resolveDir,
+                args.kind,
+              );
+            } else {
+              let packageName: string;
+              let pathParts: string[];
+              if (args.path.startsWith("@")) {
+                const [scope, name, ...rest] = args.path.split("/");
+                packageName = `${scope}/${name}`;
+                pathParts = rest;
+              } else {
+                const [name, ...rest] = args.path.split("/");
+                packageName = name;
+                pathParts = rest;
+              }
+              const packageId = loaderImpl.packageIdFromNameInPackage(
+                packageName,
+                parentPackageId,
+              );
+              const resolveDir = loaderImpl.nodeModulesDirForPackage(packageId);
+              const path = [packageName, ...pathParts].join("/");
+              return resolveInNodeModules(
+                path,
+                parentPackageId,
+                resolveDir,
+                args.kind,
+              );
+            }
+          } else {
+            throw new Error(
+              `To use "npm:" specifiers, you must specify a "cwd" and "nodeModulesDir: true", or use "loader: native".`,
+            );
+          }
+        }
         const specifier = esbuildResolutionToURL(args);
 
         // Once we have an absolute path, let the loader resolver figure out
@@ -136,21 +243,51 @@ export function denoLoaderPlugin(
             const { specifier } = res;
             return urlToEsbuildResolution(specifier);
           }
+          case "npm": {
+            let resolveDir: string;
+            if (nodeModulesDir) {
+              resolveDir = nodeModulesDir;
+            } else if (loaderImpl.nodeModulesDirForPackage) {
+              resolveDir = loaderImpl.nodeModulesDirForPackage(res.packageId);
+            } else {
+              throw new Error(
+                `To use "npm:" specifiers, you must specify a "cwd" and "nodeModulesDir: true", or use "loader: native".`,
+              );
+            }
+            const path = `${res.packageName}${res.path ?? ""}`;
+            return resolveInNodeModules(
+              path,
+              res.packageId,
+              resolveDir,
+              args.kind,
+            );
+          }
+          case "node": {
+            return {
+              path: res.path,
+              external: true,
+            };
+          }
         }
       }
       build.onResolve({ filter: /.*/, namespace: "file" }, onResolve);
       build.onResolve({ filter: /.*/, namespace: "http" }, onResolve);
       build.onResolve({ filter: /.*/, namespace: "https" }, onResolve);
       build.onResolve({ filter: /.*/, namespace: "data" }, onResolve);
+      build.onResolve({ filter: /.*/, namespace: "npm" }, onResolve);
+      build.onResolve({ filter: /.*/, namespace: "node" }, onResolve);
 
-      function onLoad(
+      async function onLoad(
         args: esbuild.OnLoadArgs,
       ): Promise<esbuild.OnLoadResult | null> {
+        if (args.namespace === "file" && args.pluginData === IN_NODE_MODULES) {
+          const contents = await Deno.readFile(args.path);
+          return { loader: "js", contents };
+        }
         const specifier = esbuildResolutionToURL(args);
         return loaderImpl.loadEsm(specifier);
       }
       // TODO(lucacasonato): once https://github.com/evanw/esbuild/pull/2968 is fixed, remove the catch all "file" handler
-      // build.onLoad({ filter: /.*\.json/, namespace: "file" }, onLoad);
       build.onLoad({ filter: /.*/, namespace: "file" }, onLoad);
       build.onLoad({ filter: /.*/, namespace: "http" }, onLoad);
       build.onLoad({ filter: /.*/, namespace: "https" }, onLoad);

--- a/src/plugin_deno_resolver.ts
+++ b/src/plugin_deno_resolver.ts
@@ -28,6 +28,9 @@ export interface DenoResolverPluginOptions {
   importMapURL?: string;
 }
 
+export const IN_NODE_MODULES = Symbol("IN_NODE_MODULES");
+export const IN_NODE_MODULES_RESOLVED = Symbol("IN_NODE_MODULES_RESOLVED");
+
 /**
  * The Deno resolver plugin performs relative->absolute specifier resolution
  * and import map resolution.
@@ -42,8 +45,11 @@ export function denoResolverPlugin(
     name: "deno-resolver",
     setup(build) {
       let importMap: ImportMap | null = null;
+      let nodeModulesPaths: Set<string>;
 
       build.onStart(async function onStart() {
+        nodeModulesPaths = new Set<string>();
+
         let importMapURL: string | undefined;
 
         // If no import map URL is specified, and a config is specified, we try
@@ -81,6 +87,24 @@ export function denoResolverPlugin(
       });
 
       build.onResolve({ filter: /.*/ }, async function onResolve(args) {
+        // If this is a node_modules internal resolution, just pass it through.
+        // Internal resolution is detected by either the "IN_NODE_MODULES" flag
+        // being set on the resolve args through the pluginData field, or by
+        // the importer being in the nodeModulesPaths set.
+        if (args.pluginData === IN_NODE_MODULES_RESOLVED) return {};
+        if (args.pluginData === IN_NODE_MODULES) return undefined;
+        if (nodeModulesPaths.has(args.importer)) {
+          const res = await build.resolve(args.path, {
+            importer: args.importer,
+            namespace: args.namespace,
+            kind: args.kind,
+            resolveDir: args.resolveDir,
+            pluginData: IN_NODE_MODULES,
+          });
+          if (!res.external) nodeModulesPaths.add(res.path);
+          return res;
+        }
+
         // The first pass resolver performs synchronous resolution. This
         // includes relative to absolute specifier resolution and import map
         // resolution.
@@ -108,7 +132,7 @@ export function denoResolverPlugin(
           const res = resolveModuleSpecifier(
             args.path,
             importMap,
-            new URL(referrer) || undefined,
+            new URL(referrer),
           );
           resolved = new URL(res);
         } else {
@@ -119,10 +143,12 @@ export function denoResolverPlugin(
         // pass. Now plugins can perform any resolution they want on the fully
         // resolved specifier.
         const { path, namespace } = urlToEsbuildResolution(resolved);
-        return await build.resolve(path, {
+        const res = await build.resolve(path, {
           namespace,
           kind: args.kind,
         });
+        if (res.pluginData === IN_NODE_MODULES) nodeModulesPaths.add(res.path);
+        return res;
       });
     },
   };

--- a/src/shared.ts
+++ b/src/shared.ts
@@ -4,13 +4,31 @@ import { MediaType } from "./deno.ts";
 export interface Loader {
   resolve(specifier: URL): Promise<LoaderResolution>;
   loadEsm(specifier: URL): Promise<esbuild.OnLoadResult>;
+
+  packageIdFromNameInPackage?(name: string, parentPackageId: string): string;
+  nodeModulesDirForPackage?(npmPackageId?: string): string;
 }
 
-export type LoaderResolution = LoaderResolutionEsm;
+export type LoaderResolution =
+  | LoaderResolutionEsm
+  | LoaderResolutionNpm
+  | LoaderResolutionNode;
 
 export interface LoaderResolutionEsm {
   kind: "esm";
   specifier: URL;
+}
+
+export interface LoaderResolutionNpm {
+  kind: "npm";
+  packageId: string;
+  packageName: string;
+  path: string;
+}
+
+export interface LoaderResolutionNode {
+  kind: "node";
+  path: string;
 }
 
 export function mediaTypeToLoader(mediaType: MediaType): esbuild.Loader {
@@ -222,4 +240,48 @@ function mediaTypeFromSpecifier(specifier: URL): MediaType {
     default:
       return "Unknown";
   }
+}
+
+export interface NpmSpecifier {
+  name: string;
+  version: string | null;
+  path: string | null;
+}
+
+export function parseNpmSpecifier(specifier: URL): NpmSpecifier {
+  if (specifier.protocol !== "npm:") throw new Error("Invalid npm specifier");
+  const path = specifier.pathname;
+  const startIndex = path[0] === "/" ? 1 : 0;
+  let pathStartIndex;
+  let versionStartIndex;
+  if (path[startIndex] === "@") {
+    const firstSlash = path.indexOf("/", startIndex);
+    if (firstSlash === -1) {
+      throw new Error(`Invalid npm specifier: ${specifier}`);
+    }
+    pathStartIndex = path.indexOf("/", firstSlash + 1);
+    versionStartIndex = path.indexOf("@", firstSlash + 1);
+  } else {
+    pathStartIndex = path.indexOf("/", startIndex);
+    versionStartIndex = path.indexOf("@", startIndex);
+  }
+
+  if (pathStartIndex === -1) pathStartIndex = path.length;
+  if (versionStartIndex === -1) versionStartIndex = path.length;
+
+  if (versionStartIndex > pathStartIndex) {
+    versionStartIndex = pathStartIndex;
+  }
+
+  if (startIndex === versionStartIndex) {
+    throw new Error(`Invalid npm specifier: ${specifier}`);
+  }
+
+  return {
+    name: path.slice(startIndex, versionStartIndex),
+    version: versionStartIndex === pathStartIndex
+      ? null
+      : path.slice(versionStartIndex + 1, pathStartIndex),
+    path: pathStartIndex === path.length ? null : path.slice(pathStartIndex),
+  };
 }

--- a/src/shared_test.ts
+++ b/src/shared_test.ts
@@ -1,0 +1,113 @@
+import { NpmSpecifier, parseNpmSpecifier } from "./shared.ts";
+import { assertEquals, assertThrows } from "../test_deps.ts";
+
+interface NpmSpecifierTestCase extends NpmSpecifier {
+  specifier: string;
+}
+
+const NPM_SPECIFIER_VALID: Array<NpmSpecifierTestCase> = [
+  {
+    specifier: "npm:package@1.2.3/test",
+    name: "package",
+    version: "1.2.3",
+    path: "/test",
+  },
+  {
+    specifier: "npm:package@1.2.3",
+    name: "package",
+    version: "1.2.3",
+    path: null,
+  },
+  {
+    specifier: "npm:@package/test",
+    name: "@package/test",
+    version: null,
+    path: null,
+  },
+  {
+    specifier: "npm:@package/test@1",
+    name: "@package/test",
+    version: "1",
+    path: null,
+  },
+  {
+    specifier: "npm:@package/test@~1.1/sub_path",
+    name: "@package/test",
+    version: "~1.1",
+    path: "/sub_path",
+  },
+  {
+    specifier: "npm:@package/test/sub_path",
+    name: "@package/test",
+    version: null,
+    path: "/sub_path",
+  },
+  {
+    specifier: "npm:test",
+    name: "test",
+    version: null,
+    path: null,
+  },
+  {
+    specifier: "npm:test@^1.2",
+    name: "test",
+    version: "^1.2",
+    path: null,
+  },
+  {
+    specifier: "npm:test@~1.1/sub_path",
+    name: "test",
+    version: "~1.1",
+    path: "/sub_path",
+  },
+  {
+    specifier: "npm:@package/test/sub_path",
+    name: "@package/test",
+    version: null,
+    path: "/sub_path",
+  },
+  {
+    specifier: "npm:/@package/test/sub_path",
+    name: "@package/test",
+    version: null,
+    path: "/sub_path",
+  },
+  {
+    specifier: "npm:/test",
+    name: "test",
+    version: null,
+    path: null,
+  },
+  {
+    specifier: "npm:/test/",
+    name: "test",
+    version: null,
+    path: "/",
+  },
+];
+
+for (const test of NPM_SPECIFIER_VALID) {
+  Deno.test(`parseNpmSpecifier: ${test.specifier}`, () => {
+    const parsed = parseNpmSpecifier(new URL(test.specifier));
+    assertEquals(parsed, {
+      name: test.name,
+      version: test.version,
+      path: test.path,
+    });
+  });
+}
+
+const NPM_SPECIFIER_INVALID = [
+  "npm:@package",
+  "npm:/",
+  "npm://test",
+];
+for (const specifier of NPM_SPECIFIER_INVALID) {
+  Deno.test(`parseNpmSpecifier: ${specifier}`, () => {
+    assertThrows(
+      () => parseNpmSpecifier(new URL(specifier)),
+      Error,
+      "Invalid npm specifier",
+    );
+  });
+}

--- a/testdata/npm/preact.tsx
+++ b/testdata/npm/preact.tsx
@@ -1,0 +1,5 @@
+/** @jsxRuntime automatic */
+/** @jsxImportSource npm:preact */
+import { renderToString } from "npm:preact-render-to-string";
+
+export default renderToString(<div>hello world</div>);

--- a/testdata/npm/react.tsx
+++ b/testdata/npm/react.tsx
@@ -1,0 +1,5 @@
+/** @jsxRuntime automatic */
+/** @jsxImportSource npm:react */
+import { renderToString } from "npm:react-dom/server";
+
+export default renderToString(<div>hello world</div>);

